### DISCUSSION TITLE
Define order in which atoms' stored value is got | Show "loading" until stored values are got on static site (SSG)

### DISCUSSION THREAD

**Discussion by valentincostam** [pmndrs/jotai#1432] [Q&A]
Hi there, 👋  

I thought to create two separate discussions but, since both questions are related to the use of `atomWithStorage` on [the same app](https://codesandbox.io/s/jotai-dependant-task-with-storage-q85lsu), I decided to keep them in one.

I hope this is ok. If not, please tell me, and I move one of them to another discussion. 🙏

> ✋ **EDIT:** I explained the problem better and added a minimal reproducible example attempt [in this answer](https://github.com/pmndrs/jotai/discussions/1432#discussioncomment-3710841).

---

## 🔢 Define order in which atoms' stored value is got

**How can I get an atom's stored value before others that depend on it?**

I'm still working on the app I showed in #1401. This time, I'm using `atomWithStorage` to persist tasks' state in Local Storage. It works great.

But I added a new feature: a checkbox to make tasks' requirements needed or not. It's like a switch for "strict mode". _This flag is `true` by default_ and it's persisted with `atomWithStorage` as well.

The problem is that flag's stored value is always got _after_ the rest. So, even if it's saved as `false`, once the page loads, its value is `true` for a few milliseconds, making requirements needed and updating tasks' state.

[Maybe this gif helps to understand the problem better.](https://s5.gifyu.com/images/tasks-atom-with-storage.gif) It shows that even though the tasks of Stage 2 are left "completed", they change to "pending" after a page refresh.

You can see the app and its code in [this sandbox](https://codesandbox.io/s/jotai-dependant-task-with-storage-q85lsu).

**How can I define the order in which values stored in Local Storage are got?**

My guess is that I need to manually set and get the values in Local Storage through one or more derived atoms. But after a couple of failed attempts, I'm not sure if that's even possible.

---

## ⌛ Show "loading" until stored values are got on static site (SSG)

**How can I set an `isLoading` variable to `true` until stored values are got from Local Storage?**

[In my app](https://codesandbox.io/s/jotai-dependant-task-with-storage-q85lsu), tasks are always the same, so I thought it would be a good use case for SSG. I chose Next.js for it.

Tasks' state is persisted in Local Storage. After I run into the mismatch error, I added the `delayInit: true` option to `atomWithStorage`, and it works.

But now there is a brief time in which tasks show their initial state before changing to their saved state.

**Is there a way to show "loading..." during that brief time?** Could you please share a simple example?

Again, I suppose I could achieve this by manually setting and getting stored values with derived atoms. But I'm not sure.

By the way, I already saw this in [`atomWithStorage`'s docs](https://jotai.org/docs/utils/atom-with-storage):

> The suggested workaround for this issue is to only render the content dependent on the storedValue client-side by wrapping it in a [custom `<ClientOnly>` wrapper](https://www.joshwcomeau.com/react/the-perils-of-rehydration/#abstractions), which only renders after rehydration.

And I also read this comment by @dai-shi in #910:

> tbh, I'm not sure `delayInit` is a good solution. It makes double renders, which is less performant. I think disabling SSR for components that depend on client data (localStorage) would be more appropriate.

The thing is, if I wrap every component dependent on stored values in `<ClientOnly>`, almost everything is rendered client-side.

So, I thought about dropping the SSG idea and just going only with client-side rendering (CSR). _But the content is always the same._ It makes total sense to me to use SSG.

**I would love to read your thoughts about this SSG vs. CSR matter.**

---

I know this is a lot. I hope it's clear and I didn't bore you.

Thanks in advance. 🙏

**Comment by dai-shi** [comment]
> 🔢 Define order in which atoms' stored value is got

> The problem is that flag's stored value is always got after the rest. 

Is it just because `delayInit` is `true`?

**Reply by valentincostam** [reply]
Thanks for quick reply, Daishi.

No, I think it's not just because `delayInit` is `true`. Tasks' state use `atomWithStorage` with that option as well. Initialization is delayed for all those atoms to prevent mismatch errors.

From the [atoms.js file](https://codesandbox.io/s/jotai-dependant-task-with-storage-q85lsu?file=/atoms.js):

```js
export const areRequirementsNeededAtom = atomWithStorage(
  `areRequirementsNeeded`,
  true,
  {
    ...createJSONStorage(() => localStorage),
    delayInit: true
  }
);

export const tasksAtom = atom(
  tasks.map((task) => ({
    ...task,
    state: atomWithStorage(`task-${task.id}-state`, "pending", {
      ...createJSONStorage(() => localStorage),
      delayInit: true
    }),
    isEnabled: createAtomIsEnabledFromTaskId(task.id)
  }))
);
```

Hope that code makes it clearer. Let me know if not.

Thanks again. 🙏

**Reply by dai-shi** [reply]
I mean, if you remove `delayInit: true`, the value is the expected one in the first render, right? The hydration mismatch error (which is a warning?) is another beast.

**Reply by valentincostam** [reply]
I think so.

I removed `delayInit: true` from those atoms. It looks like it's the expected value but still, the problem persists.

If it helps, [here is a gif showing the inconsistent state after reloading the page](https://s5.gifyu.com/images/atoms-without-delay-init.gif). With the checkbox unchecked, the last task should remain enabled and "completed" after the page refresh, but it is not. In the console, you can see the hydration mismatch error.

Let me know what else I can do to help you to understand the problem better.

Thanks for taking the time. 💪

**Comment by valentincostam** [comment]
Hi everyone,

I tried to make a minimal reproducible example. At first, the problem occurred over and over. Then, suddenly, it stopped, without any change in the code. Weird.

Anyway, I share the example because I think it helps to understand the problem better.

In the example, there are two fields, a `checkbox` and a `select` with two options. Both fields' value is stored in Local Storage using `atomWithStorage`.

To select "Option 2", the checkbox must be unchecked.

If you choose "Option 2" and reload the page, it's likely that it doesn't remain the selected one. Even when the checkbox is not checked.

I think the reason for this behavior is that sometimes `select`'s stored value is got _before_ `checkbox`'s one. In that case, for a brief time (milliseconds), the `checkbox` is checked, so the `select` cannot be set to "Option 2".

**How can I make sure `checkbox`'s stored value is _always_ got first?**

[Here is the sandbox.](https://codesandbox.io/s/dependent-atoms-with-storage-cyilt1)

Since it's not that much code, I paste it here:

```js
const isCheckedAtom = atomWithStorage("isChecked", true);
const selectedOptionAtom = atomWithStorage("selectedOption", "option-1");

export default function App() {
  const [isChecked, setIsChecked] = useAtom(isCheckedAtom);
  const [selectedOption, setSelectedOption] = useAtom(selectedOptionAtom);

  useEffect(() => {
    if (isChecked && selectedOption === "option-2") {
      setSelectedOption("option-1");
    }
  }, [setSelectedOption, isChecked, selectedOption]);

  return (
    <>
      <input
        type="checkbox"
        id="is-checked"
        checked={isChecked}
        onChange={() => setIsChecked(!isChecked)}
      />
      <label htmlFor="is-checked">Is checked? {isChecked ? "Yes" : "No"}</label>

      <br />

      <select
        value={selectedOption}
        onChange={(e) => setSelectedOption(e.target.value)}
      >
        <option value="option-1">Option 1</option>
        <option value="option-2" disabled={isChecked}>
          Option 2 (requires checkbox unchecked)
        </option>
      </select>
    </>
  );
}
```

Thanks for reading.

**Reply by dai-shi** [reply]
Oh, great! You reproduced the issue with codesandbox (CRA).
Unfortunately, the issue isn't reproduced in my chrome. But, I see the issue is probably due to the use of useEffect. tldr is useEffect isn't the right tool for deriving state (in general, not just with jotai).

So, if that causes an issue, I would probably try to avoid useEffect. Luckily, jotai can derive state.

```jsx
import { atom, useAtom } from "jotai";
import { atomWithStorage } from "jotai/utils";

const storedIsCheckedAtom = atomWithStorage("isChecked", true);
const storedSelectedOptionAtom = atomWithStorage("selectedOption", "option-1");

const isCheckedAtom = atom(
  (get) => get(storedIsCheckedAtom),
  (get, set, isChecked) => {
    set(storedIsCheckedAtom, isChecked)
    if (isChecked) {
      set(storedSelectedOptionAtom, 'option-1');
    }
  }
);
const selectedOptionAtom = atom(
  (get) => get(storedIsCheckedAtom) ? "option-1" : get(storedSelectedOptionAtom),
  (get, set, selectedOption) => {
    const isChecked = get(storedIsCheckedAtom);
    set(storedSelectedOptionAtom, isChecked ? "option-1" : selectedOption);
  }
);

export default function App() {
  const [isChecked, setIsChecked] = useAtom(isCheckedAtom);
  const [selectedOption, setSelectedOption] = useAtom(selectedOptionAtom);

  return (
    <>
      <input
        type="checkbox"
        id="is-checked"
        checked={isChecked}
        onChange={() => setIsChecked(!isChecked)}
      />
      <label htmlFor="is-checked">Is checked? {isChecked ? "Yes" : "No"}</label>

      <br />

      <select
        value={selectedOption}
        onChange={(e) => setSelectedOption(e.target.value)}
      >
        <option value="option-1">Option 1</option>
        <option value="option-2" disabled={isChecked}>
          Option 2 (requires checkbox unchecked)
        </option>
      </select>
    </>
  );
}
```

It's a bit of ceremony, but it's declarative, so it shouldn't cause any issue.

**Reply by valentincostam** [reply]
Thank you so much for taking the time again, Daishi. I appreciate it.

Now I know how to derive state correctly.

That said, I think I oversimplified the example. I had a hard time trying to replicate your approach on my app.

In the end, I made it work. The code looks kind of confusing, though. [This is the sandbox](https://codesandbox.io/s/jotai-dependant-task-with-storage-forked-mvrro1?file=/atoms.js) in case someone wants to take a look at it. Any feedback is welcome.

One last question to deem the discussion as closed: **Is there any way to set an `isLoading` atom to `true` when another atom's stored value wasn't got yet?**

Since I'm using SSG and `atomWithStorage` with the `delayInit: true` option, I would like to show a "Loading..." until the stored value is got.

I was thinking about something like this:

```js
const isLoadingAtom = atom((get) => typeof get(storedIsCheckedAtom) === 'undefined');
```

Is it possible?

Thanks. 🙏

**Reply by dai-shi** [reply]
That would require a trick. One out of my head is have a special value as default. And check it on mount. Either with useEffect or onMount. We may need a little delay. Otherwise, you may need a custom storage for atomWithStorage. Sorry for a rough suggestion. I can provide a code snippet if you like any of the approaches.

**Reply by valentincostam** [reply]
Thank you, Daishi.

Do you mean something like this?

```js
export default function App() {
  const [selectedOption, setSelectedOption] = useAtom(selectedOptionAtom);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (selectedOption !== "default-option") {
      setIsLoading(false);
    }
  }, [selectedOption]);

  return (
    <>
      {isLoading && <p>Loading...</p>}
      // ...
    </>
  );
}
```

Even though `isLoading` is not an atom, is it not wrong deriving state with `useEffect` in this case?

If you can provide a code snippet, it would be great. 💪

**Reply by dai-shi** [reply]
Here are some ideas. I didn't try them. 

```js
const EMPTY = Symbol()
const initialValue = null
const a = atomWithStorage('key', EMPTY)
a.onMount = (set) => {
  set(prev => prev === EMPTY ? initialValue : EMPTY)
}
```


```js
const storage = {
  getItem(key) {
    const s = localStorage.getItem(key)
    if (s === null) {
      return initialValue 
    }
    return JSON.parse(s)
  },
  // ...
}
const a = atomWithStorage('key', EMPTY, () => storage)
```

**Reply by quoctienkt** [reply]
I don't like the way manually handling isLoading, it should be built-in like React Query or RTK Query. If jotai cannot support isLoading, don't use asynchronous, also the main reason I left using jotai and switch to redux, thanks
